cpp/src/Ice/ConnectionI.h

```diff
@@ -307,13 +307,20 @@ namespace Ice
         IceInternal::SocketOperation read(IceInternal::Buffer&);
         IceInternal::SocketOperation write(IceInternal::Buffer&);
 
-        // A connection is at rest if it is active and has no outstanding invocations or dispatches.
+        // A connection is at rest if:
+        // - it is active and has no outstanding invocations or dispatches.
+        // - it is not reading or writing a message other than a connection validation message
         // We schedule the inactivity timer task when it enters the "at rest" state, and we cancel this timer task when
-        // the connection is about to leave this state.
+        // the connection is about to leave this state. The writing or reading of a connection validation message isn't
+        // considered as application "activity" since it's used internally to check the connection liveliness.
         // Must be called with _mutex locked.
         bool isAtRest() const noexcept
         {
-            return _state == StateActive && _dispatchCount == 0 && _asyncRequests.empty();
+            return _state == StateActive && _dispatchCount == 0 && _asyncRequests.empty() &&
+                   (_isExpectingMessageHeaderOnNextRead ||
+                    static_cast<uint8_t>(_readStream.b[8]) == IceInternal::validateConnectionMsg) &&
+                   (_writeStream.b.empty() || _writeStream.i == _writeStream.b.end() ||
```

Comment:
On the write side, should we examine all queued messages and not just the current one?

For example, we could be writing a ValidateConnection message while a Request message is queued. Naturally, this would add complexity.

Reply:
The reason for this complex check is: when I decrement the `_dispatchCount` to 0, I could be reading a new Request or writing a Reply (really waiting for the next bytes), and I don't want to schedule the inactivity timer at this point.

Reply:
```cpp
(_isExpectingMessageHeaderOnNextRead ||
                    static_cast<uint8_t>(_readStream.b[8]) == IceInternal::validateConnectionMsg)
```

I understand the first part: we're waiting for a new message. But when can be encounter the second part?
A ValidateConnection message is 14 bytes, so immediately after we read it, we don't set `_isExpectingMessageHeaderOnNextRead` to true to read the next message? 

When do we unlock the connection mutex with _isExpectingMessageHeaderOnNextRead = false and a ValidateConnection in the _readStream?

Reply:
> On the write side, should we examine all queued messages and not just the current one?

Indeed, we should check if there are queued message other than the connection validation message. I've changed the code to:
- never queue connection validation messages in `sendMessage`
- check if there are queued messages and the message being written is not a connection validation message for the isAtRest check

> But when can we encounter the second part?

You're right we can't, I removed it.

> When do we unlock the connection mutex with _isExpectingMessageHeaderOnNextRead = false and a ValidateConnection in the _readStream?

When the mutex is unlocked, this condition is never true. In other words, when the mutex is unlocked in `message` either:
- `_isExpectingMessageHeaderOnNextRead==false` and `_readStream` doesn't contain a connection validation message
- `_isExpectedMessageHeaderOnNextRead==true` and `_readStream` is considered empty